### PR TITLE
ids serialization and deserialization:

### PR DESCRIPTION
**Pull request by mjkkirschner**
### Purpose

This PR deals with both https://jira.autodesk.com/browse/QNTM-692 and https://jira.autodesk.com/browse/QNTM-700

Leaves `workspaceModel`'s `Guid` property serializing to `Uuid` as it is today.
It changes `AnnotationModel` `PortModel` and `ConnectorModel` and `NodeModel` to serialize their `Guid` property as `Id` to be more explicit about that fact that these properties do not **need** to be globally unique but have different uniqueness requirements that are explained in their summaries.

When `id` properties are deserialized they are parsed as if they are guids, if they are not valid guids, we generate a deterministic guid based on the id string.

The element resolver still uses guids to remap the model as before. I've also added some checks to it that throw exceptions if a duplicate key is added to the resolve map or resolver models dict - all ids should at a minimum be unique within the graph so this should be a valid check to do.

- [x] I need to add tests which create some non guid ids and validate the graphs deserialize and run correctly.

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@QilongTang 

### FYIs

@ramramps @ikeough @gregmarr

### CODE REVIEW

File: src/DynamoCore/Graph/Connectors/ConnectorModel.cs [hunk 1]
```diff
@@ -39,6 +40,22 @@ public class ConnectorModel : ModelBase
         /// </summary>
         public PortModel End { get; private set; }
 
+        /// <summary>
+        /// ID of the Connector, which is unique on in the graph.
```
Comment (QilongTang) [inline src/DynamoCore/Graph/Connectors/ConnectorModel.cs]: `within the graph`?

**Comment by mjkkirschner** [comment]
I've added a new set of tests based on the serialization tests that modify the generated json to have non guid ids - these are then deserialized, run and and compared.

I'll fix up the merge conflicts and this should be ready for review.

File: src/Workspaces/SerializationConverters.cs [hunk 2]
```diff
@@ -54,10 +54,19 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
             var obj = JObject.Load(reader);
             var type = Type.GetType(obj["$type"].Value<string>());
-            
-            var guid = Guid.Parse(obj["Uuid"].Value<string>());
+
+            //if the id is not a guid, makes a guid based on the id of the node
+            Guid nodeId;
+            if (!Guid.TryParse((obj["Id"].Value<string>()), out nodeId))
+            {
+                nodeId = GuidUtility.Create(GuidUtility.UrlNamespace, (obj["Id"].Value<string>()));
```
Comment (gregmarr) [inline src/Workspaces/SerializationConverters.cs]: This pattern occurs multiple times, seems like it should become a helper function.

Comment (mjkkirschner) [inline src/Workspaces/SerializationConverters.cs]: good catch- thanks, should be straight forward.

File: src/DynamoCore/Graph/Workspaces/Serialiaztion/SerializationConverters.cs [hunk 3]
```diff
@@ -52,10 +52,14 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
 
             var obj = JObject.Load(reader);
             var type = Type.GetType(obj["$type"].Value<string>());
-            
-            var guid = Guid.Parse(obj["Uuid"].Value<string>());
+
+            //if the id is not a guid, makes a guid based on the id of the node
+            Guid nodeId = GuidUtility.tryParseOrCreateGuid(obj["Id"].Value<string>());
```
Comment (QilongTang) [inline src/DynamoCore/Graph/Workspaces/Serialiaztion/SerializationConverters.cs]: Let's just assign this to `guid` directly

**Comment by mjkkirschner** [comment]
@QilongTang addressed the comment.

File: test/DynamoCoreTests/SerializationTests.cs [hunk 4]
```diff
@@ -260,7 +335,7 @@ private void DoWorkspaceOpenAndCompare(string filePath)
 
             ConvertCurrentWorkspaceToDesignScriptAndSave(filePathBase);
 
-            string json = ConvertCurrentWorkspaceToJsonAndSave(model, filePathBase);
+            string json = saveFunction(model, filePathBase);
```
Comment (QilongTang) [inline test/DynamoCoreTests/SerializationTests.cs]: Neat

**Comment by QilongTang** [comment]
@mjkkirschner LGTM

**Pull request merged**